check/check.go

```diff
@@ -77,37 +92,140 @@ type Group struct {
 
 // Run executes the audit commands specified in a check and outputs
 // the results.
-func (c *Check) Run() {
+func (c *Check) Run(definedConstraints map[string][]string) {
+	var subCheck *BaseCheck
+	if c.SubChecks == nil {
+		subCheck = &BaseCheck{
+			Commands:    c.Commands,
+			Tests:       c.Tests,
+			Type:        c.Type,
+			Audit:       c.Audit,
+			Remediation: c.Remediation,
+		}
+	} else {
+		subCheck = getFirstValidSubCheck(c.SubChecks, definedConstraints)
 
-	if c.Type == "skip" {
-		c.State = INFO
+		if subCheck == nil {
+			c.State = WARN
+			glog.V(1).Info("Failed to find a valid sub check, check ", c.ID)
+			return
+		}
+	}
+
+	var out bytes.Buffer
+	var errmsgs string
+
+	c.State = runAuditCommands(*subCheck, &out, &errmsgs)
+
+	if c.State != "" {
 		return
 	}
 
-	// If check type is manual or the check is not scored, force result to WARN
-	if c.Type == "manual" || !c.Scored {
+	if errmsgs != "" {
+		glog.V(2).Info(errmsgs)
+	}
+
+	finalOutput := subCheck.Tests.Execute(out.String())
+	if finalOutput != nil {
+		c.ActualValue = finalOutput.ActualResult
+		c.ExpectedResult = finalOutput.ExpectedResult
+
+		if finalOutput.TestResult {
+			c.State = PASS
+		} else {
+			c.State = FAIL
+		}
+	} else {
 		c.State = WARN
+		glog.V(1).Info("Test output contains a nil value")
 		return
 	}
+}
 
-	var out bytes.Buffer
-	var errmsgs string
+// textToCommand transforms an input text representation of commands to be
+// run into a slice of commands.
+// TODO: Make this more robust.
+func textToCommand(s string) []*exec.Cmd {
+	cmds := []*exec.Cmd{}
+
+	cp := strings.Split(s, "|")
+
+	for _, v := range cp {
+		v = strings.Trim(v, " ")
+
+		// TODO:
+		// GOAL: To split input text into arguments for exec.Cmd.
+		//
+		// CHALLENGE: The input text may contain quoted strings that
+		// must be passed as a unit to exec.Cmd.
+		// eg. bash -c 'foo bar'
+		// 'foo bar' must be passed as unit to exec.Cmd if not the command
+		// will fail when it is executed.
+		// eg. exec.Cmd("bash", "-c", "foo bar")
+		//
+		// PROBLEM: Current solution assumes the grouped string will always
+		// be at the end of the input text.
+		re := regexp.MustCompile(`^(.*)(['"].*['"])$`)
+		grps := re.FindStringSubmatch(v)
+
+		var cs []string
+		if len(grps) > 0 {
+			s := strings.Trim(grps[1], " ")
+			cs = strings.Split(s, " ")
+
+			s1 := grps[len(grps)-1]
+			s1 = strings.Trim(s1, "'\"")
+
+			cs = append(cs, s1)
+		} else {
+			cs = strings.Split(v, " ")
+		}
+
+		cmd := exec.Command(cs[0], cs[1:]...)
+		cmds = append(cmds, cmd)
+	}
+
+	return cmds
+}
+
+func isShellCommand(s string) bool {
+	cmd := exec.Command("/bin/sh", "-c", "command -v "+s)
+
+	out, err := cmd.Output()
+	if err != nil {
+		return false
+	}
+
+	if strings.Contains(string(out), s) {
+		return true
+	}
+	return false
+}
+
+func runAuditCommands(c BaseCheck, out *bytes.Buffer, errmsgs *string) State {
```

Comment:
Since  Go gives us the luxury of multiple return values, it's more idiomatic to use them. Here you're really returning three variables (the state, `out`, and `errmsgs`) and it's easier to read if you pass them back that way. You also wouldn't need to pass as pointers. In cases like this where performance isn't really an issue, I'd be inclined not to use pointers as it's easier to read.